### PR TITLE
Low fidelity implementation of synthetic storage to list handles from a serialized arc

### PR DESCRIPTION
**Pull request by mykmartin**

### CODE REVIEW

File: runtime/ts/storage/firebase-storage.ts [hunk 1]
```diff
@@ -147,11 +147,8 @@ baseStorageKey(type: Type, keyString: string): string {
     return new FirebaseKey(s);
   }
 
-  // referenceMode is only referred to if shouldExist is false, or if shouldExist is 'unknown'
-  // but this _join creates the storage location. 
-  async _join(id: string, type: Type, keyString: string, shouldExist: boolean | 'unknown', referenceMode = false) {
-    assert(!type.isVariable);
-    assert(!type.isCollection || !type.primitiveType().isVariable);
+  // Exposed for SyntheticStorage. TODO: better API
+  attach(keyString: string) : {fbKey: FirebaseKey, reference: firebase.database.Reference} {
```
Comment (shans) [inline runtime/ts/storage/firebase-storage.ts]: Add a comment that this is only here to share the this.apps list and that we will eventually refactor arc storage such that it's sufficiently variable-like that we can access the manifest using a standard FirebaseVariable.

File: runtime/ts/storage/synthetic-storage.ts [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    super(arcId);
+    this.firebaseStorage = firebaseStorage;
+  }
+
+  async construct(id: string, type: Type, keyFragment: string) : Promise<SyntheticCollection> {
+    throw new Error('cannot construct synthetic storage providers; use connect');
+  }
+
+  async connect(id: string, type: Type, key: string) : Promise<SyntheticCollection> {
+    // TODO: fix the API so we don't need this workaround
```
Comment (shans) [inline runtime/ts/storage/synthetic-storage.ts]: instead: // TODO: add handle type to type system.

**Comment by shans** [comment]
@sjmiles 

Usage is something along the lines of:

```
let storageFactory = arc._storageFactory;
let fbKey = <arc's firebase location>
let handleStore = await storageFactory.connect('some-id', null, 'synthetic://arc/handles/${fbKey}');
let handles = await handleStore.toList();
```

For any handle in handles, you can then:

```
let store = await storageFactory.connect('some-other-id', handle.type, handle.storageKey);
// await store.toList() or store.get() or whatever 
```

Also you'll have a `handle.tags` list to play with.

File: runtime/ts/storage/firebase-storage.ts (hunk 1 above)
Comment (mykmartin) [reply runtime/ts/storage/firebase-storage.ts]: Done.

File: runtime/ts/storage/synthetic-storage.ts (hunk 2 above)
Comment (mykmartin) [reply runtime/ts/storage/synthetic-storage.ts]: Done.

**Comment by mykmartin** [comment]
Small clarification to Shane's example: the synthetic storage expects the "arc's firebase location" bit to refer directly to the serialized arc data; e.g. ```firebase://.../0_4_1-alpha/arcs/-LLSeBXnEplPcQP_XjbN/serialization```

If you'd prefer to use the more standard arc storageKey and have SyntheticStorage look for the serialization child, that's both easy and fine.

**Pull request merged**